KAFKA-12432: AdminClient should time out nodes that are never ready

cmccabe opened the pull request:
Previously, if we assigned one or more calls to a remote node, but it
never became available, we would block until the calls hit their the API
timeout.  This was particularly unfortunate in the case where the calls
could have been sent to a different node in the cluster.  This PR fixes
this behavior by timing out pending connections to remote nodes if they
take longer than the request timeout.

There are a few other small cleanups in this PR: it removes the
unecessary Call#aborted, sets Call#curNode to null after the call has
failed to avoid confusion when debugging or logging, and adds a
"closing" boolean rather than setting newCalls to null when the client
is closed.  Also, it increases the log level of the log message that
indicates that we timed out some calls because AdminClient closed.

hachikuji commented:
> Previously, we relied solely on the broker to enforce
request.timeout.ms. If the broker failed to enforce this timeout, the
admin client would hang until the call's API timeout hit, which would
return an error to the caller.

I don't think this is accurate. The `NetworkClient` enforces the request timeout, which is passed through the `ClientRequest` object. If the response is not received before the timeout expires, then `NetworkClient` will disconnect.

cmccabe committed:
KAFKA-12432: AdminClient should time out nodes that are never ready

Previously, if we assigned one or more calls to a remote node, but it
never became available, we would block until the calls hit their the API
timeout.  This was particularly unfortunate in the case where the calls
could have been sent to a different node in the cluster.  This PR fixes
this behavior by timing out pending connections to remote nodes if they
take longer than the request timeout.

There are a few other small cleanups in this PR: it removes the
unecessary Call#aborted, sets Call#curNode to null after the call has
failed to avoid confusion when debugging or logging, and adds a
"closing" boolean rather than setting newCalls to null when the client
is closed.  Also, it increases the log level of the log message that
indicates that we timed out some calls because AdminClient closed.

cmccabe commented:
> I don't think this is accurate. The NetworkClient enforces the request timeout, which is passed through the ClientRequest object. If the response is not received before the timeout expires, then NetworkClient will disconnect.

Thanks for the correction.  This part of the change description was inaccurate-- I have removed it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1136,17 +1185,12 @@ private void timeoutCallsInFlight(TimeoutProcessor processor) {
-                        // has been closed, the calls should be returned by the next client#poll(),
-                        // and handled at that point.
-                    }
+                    log.debug("Disconnecting from {} due to timeout while awaiting {}", nodeId, call);
```
I wonder if we can raise this to info? This is a really important event to see in the logs, and we are always left guessing about it when the user does not have debug logging enabled. The frequency at the rate of the request timeout should keep it from being too spammy.

cmccabe replied:
OK, let's raise it to `INFO`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1089,29 +1106,61 @@ private long sendEligibleCalls(long now) {
+                if (!callsInFlight.getOrDefault(node.idString(), Collections.emptyList()).isEmpty()) {
+                    log.trace("Still waiting for other calls to finish on node {}.", node);
+                    nodeReadyDeadlines.remove(node);
+                    continue;
```
Maybe I am missing something here. Why do we need to continue? Do we not support multiple in-flight requests? If not, then why does `callsInFlight` map to a list?

cmccabe replied:
Notice that we set `maxInFlightRequestsPerConnection` to 1 when constructing the `NetworkClient`.  We don't support sending multiple requests to a single node on a single connection in `AdminClient`.  I think we could add this support, but we'd have to check how the server handled it since we've never done it before.  Maybe there should be a JIRA.

Also, if we do choose to add this support for multiple outstanding requests per node per socket we'd need some way to distinguish between "waiting for a chance to use this connection" from "waiting for this connection to be opened" in NetworkClient.  Currently ready just returns a boolean, which isn't enough information to distinguish these two cases.  We could probably add a new method that returned an enum or something.

cmccabe replied:
Oh, and on the question of why inflight requests has a list, it was done for future-proofing

hachikuji replied:
Ok. I think we might as well simplify the type since we already have logic which assumes one inflight request. It doesn't feel like we're buying ourselves much in terms of future-proofing.

cmccabe replied:
ok.  I changed it to just `Map<String, Call>`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1089,29 +1106,61 @@ private long sendEligibleCalls(long now) {
+                        }
+                        pollTimeout = Math.min(pollTimeout, deadline - now);
+                    } else {
+                        nodeReadyDeadlines.put(node, now + requestTimeoutMs);
```
It seems like it would be quite intuitive to use `socket.connection.setup.timeout.ms`. Is it because of the complexity of `socket.connection.setup.timeout.max.ms` that we don't do this? Basically we don't know how long the network client itself is planning to wait for the connection to be established.

cmccabe replied:
The complexity of the min / max issue is one thing.  Another thing is that we don't know when the connection has been established, and when it has not.  NetworkClient doesn't expose this information.  `NetworkClient#ready` may return false for a variety of reasons, many of which are not indicative of connection establishment.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1089,29 +1106,61 @@ private long sendEligibleCalls(long now) {
-                Call call = calls.remove(0);
-                int requestTimeoutMs = Math.min(KafkaAdminClient.this.requestTimeoutMs,
+                int remainingRequestTime;
+                Long deadlineMs = nodeReadyDeadlines.remove(node);
```
So I guess the intent here is to count the "ready time" toward the request timeout. I think that's reasonable. It means that request timeout represents the maximum time we will commit to trying to get a response from a particular node before retrying. It might be worth a comment to explain this?

cmccabe replied:
I added a comment

cmccabe committed:
Raise "disconnecting..." log message to INFO level

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1089,29 +1106,63 @@ private long sendEligibleCalls(long now) {
+                            iter.remove();
+                            continue;
+                        }
+                        pollTimeout = Math.min(pollTimeout, deadline - now);
```
Might be a good idea to use `Math.max(0, deadline - now)` since system time is non-monotonic.

cmccabe replied:
Let's just do that in the call to `NetworkClient#poll`.  In general I agree that monotonic time would have been better (maybe this needs a rewrite...)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1089,29 +1106,63 @@ private long sendEligibleCalls(long now) {
+                    try {
+                        requestBuilder = call.createRequest(timeoutMs);
+                    } catch (Throwable throwable) {
+                        call.fail(now, new KafkaException(String.format(
```
I noticed that we lose the throwable that we caught here. Can we add it as a cause?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1367,15 +1414,15 @@ private void processRequests() {
             if (call.tries > maxRetries) {
                 log.debug("Max retries {} for {} reached", maxRetries, call);
-                call.fail(time.milliseconds(), new TimeoutException());
+                call.handleTimeoutFailure(time.milliseconds(), new TimeoutException());
```
While we're here, can we set the exception message to indicate that max retries had been reached?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -5307,6 +5308,71 @@ public void testUnregisterBrokerTimeoutMaxWait() {
+            while (true) {
+                time.sleep(1);
+                try {
+                    disconnectFuture.get(1, TimeUnit.MICROSECONDS);
```
Might be simpler to check `isDone`.

cmccabe replied:
In this case, `isDone` is not supposed to be true until we advance the clock (so that a new node can be tried, after the previous one disconnected)

cmccabe committed:
Address review comments...

cmccabe commented:
For some reason the branch for this PR just stopped changing even when I pushed to it.  Maybe it's a github bug?

I opened a different PR, https://github.com/apache/kafka/pull/10395 , that has the changes we discussed here.

hachikuji reviewed (approved):
LGTM (pending Jenkins approval). Thanks for the patch!
